File: ios/Packages/GemstonePrimitives/Sources/AutocloseValidator.swift
// Copyright (c). Gem Wallet. All rights reserved.

import Foundation
import Gemstone
import Primitives

public struct AutocloseValidator: Sendable {
    private let validator: Gemstone.AutocloseValidator

    public init(type: Primitives.TpslType, direction: Primitives.PerpetualDirection, marketPrice: Double) throws {
        validator = Gemstone.AutocloseValidator(triggerType: type.map(), direction: direction.json(), marketPrice: marketPrice)
    }

    public func validate(_ price: Double) -> AutocloseValidation {
        validator.validate(price: price)
    }
}

File: android/gemcore/src/main/kotlin/com/gemwallet/android/domains/perpetual/autoclose/AutocloseValidator.kt
package com.gemwallet.android.domains.perpetual.autoclose

import com.gemwallet.android.domains.perpetual.toGem
import com.gemwallet.android.ext.toGem
import com.wallet.core.primitives.PerpetualDirection
import com.wallet.core.primitives.TpslType
import uniffi.gemstone.AutocloseValidation
import uniffi.gemstone.AutocloseValidator as GemAutocloseValidator

class AutocloseValidator(
    type: TpslType,
    direction: PerpetualDirection,
    marketPrice: Double,
) {
    private val validator = GemAutocloseValidator(type.toGem(), direction.toGem(), marketPrice)

    fun validate(price: Double?): AutocloseValidation =
        price?.let { validator.validate(it) } ?: AutocloseValidation.VALID
}
